rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+
+```json
+"http://yourdomain/namespace/property-type/type_name"
+```
+
+as the identifier (assuming that the domain decides that type names are unique across a namespace).
+
+Where it's possible to guarantee, this URI should be valid indefinitely.
+
+> ⚠️ New iterations and changes to the type must occur under new version numbers, components of the Block Protocol ecosystem will certainly implement caching and local-persistence logic, and the spec consciously makes no recommendation that these caches are updated.
+>
+> **As such, a specific version of a Type is to be treated as an immutable record**.
+
+# Reference-level explanation
+
+## What is a "version" of a Type
+
+Any published (accessible) iteration of a Type can be considered a "version" of the Type. Once a specific iteration of a type is made accessible via a URI, it should be considered immutable.
+
+- This means that any change made to the Type (for example changing the description, adding a field, etc) constitutes reason for a new version.
+- If multiple changes are made prior to making it accessible (i.e. batching the changes), then a separate version is not needed for each individual change, but they can all be 'released' together.
+
+## URI Format and Version Numbers
+
+Type URIs, and their version numbers are defined as outlined in the [Guide-level explanation](#guide-level-explanation).
+
+### Schemas
+
+1.  The `"$id"` field of a Type should be equal to the Type's _versioned URI_.
+1.  When referencing another Type using `"$ref"`, the reference should be equal to the Type's _versioned URI_.
+1.  When used as the key of a property, the _base URI_ should be used.
+
+Example:
+
+```json
+{
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/@alice/types/entity-type/book",
+  "type": "object",
+  "title": "Book",
+  "properties": {
+    "https://blockprotocol.org/@alice/types/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/types/property-type/name/v/13"
+    },
+    "https://blockprotocol.org/@alice/types/property-type/published-on": {
+      "$ref": "https://blockprotocol.org/@alice/types/property-type/published-on/v/23"
+    },
+    "https://blockprotocol.org/@alice/types/property-type/blurb": {
+      "$ref": "https://blockprotocol.org/@alice/types/property-type/blurb/v/4"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/types/property-type/name"]
+}
+```
+
+```json
+{
+  "https://blockprotocol.org/@alice/types/property-type/name": "The Time Machine",
+  "https://blockprotocol.org/@alice/types/property-type/published-on": "1895-05",
+  "https://blockprotocol.org/@alice/types/property-type/blurb": ...
+}
+```
+
+## Determining Type Compatibility
+
+A key part of the reasoning for picking this approach depends on the ability to determine compatibility between various types.
```

Comment:
Actually don't think I agree with this one, we can determine compatibility in this fashion regardless of the approach of versioning. The reasoning of not choosing semver somewhat depends on us being able to do things more dynamically because we can (mostly) determine compatibility between types.

Reply:
Fair enough